### ISSUE TITLE
very very thank your

### ISSUE THREAD

**Issue by cheapWhite** [segmentio/kafka-go#868]
Hi, I have the same err with the version v0.4.28 and 0.10.2 for Kafka server version.
The error comes from readFull also.
If I use the Writer, I have the: error unexpected EOF
```
w := &kafka.Writer{
	Addr:  kafka.TCP("kafka:9093"),
	Topic: "topic",
	Transport: &kafka.Transport{
		SASL: plain.Mechanism{
			Username: "user",
			Password: "XXXXX",
		},
		TLS: &tls.Config{},
	},
}

err = w.WriteMessages(context.Background(),
	kafka.Message{
		Value: []byte("message"),
	},
)
if err != nil {
	log.Fatal("failed to write messages:", err)
}
```

But if I use Conn, it's work.
```
partition := 0
dialer := &kafka.Dialer{
	Timeout:   10 * time.Second,
	DualStack: true,
	TLS: &tls.Config{
		InsecureSkipVerify: true,
	},
	SASLMechanism: plain.Mechanism{
		Username: "user",
		Password: "XXXXX",
	},
}

conn, err := dialer.DialLeader(context.Background(), "tcp", "kafka:9093", "topic", partition)
if err != nil {
	log.Fatal("failed to dial leader:", err)
}

conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
_, err = conn.WriteMessages(
	kafka.Message{
		Value: []byte("message"),
	},
)
if err != nil {
	log.Error("failed to write messages : " + err.Error())
}
```

With version 0.4.0 and the older versions, it works. The error seems to come with this commit: https://github.com/segmentio/kafka-go/commit/81bd03ced29aa1fdc37e54ab79ae53884e0a8c21

_Originally posted by @Jrmy2402 in https://github.com/segmentio/kafka-go/issues/795#issuecomment-1034182888_

**Comment by rhansen2** [comment]
Hi @cheapWhite, 
Thanks for reporting your issue.

Could you try out https://github.com/rhansen2/kafka-go/tree/transport-saslv0 and let me know if it helps?

Thanks!

**Comment by rhansen2** [comment]
Fixed via #869
